File: platform/lib/contract.ts
// Contract types and parser matching crates/aule-schema/src/contract.rs

import { validatePermission } from "./permissions";

// --- Types ---

export type Determinism = "deterministic" | "bounded" | "probabilistic";

export type LatencyClass = "fast" | "moderate" | "slow";

export type CostClass = "free" | "low" | "medium" | "high";

/**
 * Input/Output can be the literal string "prompt" for prompt-based skills,
 * or a JSON Schema object for structured skills.
 */
export type InputOutput = "prompt" | Record<string, unknown>;

export interface ContractError {
  code: string;
  description: string;
}

export interface BehavioralMetadata {
  latencyClass?: LatencyClass;
  costClass?: CostClass;
  sideEffects?: boolean;
}

export interface Contract {
  version: string;
  inputs: InputOutput;
  outputs: InputOutput;
  permissions: string[];
  determinism: Determinism;
  errors?: ContractError[];
  behavior?: BehavioralMetadata;
}

// --- Errors ---

export class ContractParseError extends Error {
  constructor(message: string) {
    super(message);
    this.name = "ContractParseError";
  }
}

export class ContractValidationError extends Error {
  constructor(
    message: string,
    public readonly errors: string[]
  ) {
    super(message);
    this.name = "ContractValidationError";
  }
}

// --- Validation result ---

export interface ContractValidationResult {
  valid: boolean;
  errors: string[];
  warnings: string[];
}

// --- Helpers ---

const VALID_DETERMINISM = new Set<string>(["deterministic", "bounded", "probabilistic"]);
const VALID_LATENCY = new Set<string>(["fast", "moderate", "slow"]);
const VALID_COST = new Set<string>(["free", "low", "medium", "high"]);

function isPrompt(value: unknown): value is "prompt" {
  return value === "prompt";
}

function isInputOutput(value: unknown): value is InputOutput {
  return isPrompt(value) || (typeof value === "object" && value !== null && !Array.isArray(value));
}

function isSemver(s: string): boolean {
  const parts = s.split(".");
  if (parts.length !== 3) return false;
  return parts.every((p) => /^\d+$/.test(p));
}

// --- Parsing ---

/**
 * Parse a contract from an unknown data value (e.g. inline from manifest YAML
 * or parsed from a standalone contract file).
 */
export function parseContract(data: unknown): Contract {
  if (typeof data !== "object" || data === null || Array.isArray(data)) {
    throw new ContractParseError("contract must be an object");
  }

  const obj = data as Record<string, unknown>;

  // version (required)
  if (typeof obj.version !== "string") {
    throw new ContractParseError("contract.version is required and must be a string");
  }

  // inputs (required)
  if (!("inputs" in obj)) {
    throw new ContractParseError("contract.inputs is required");
  }
  if (!isInputOutput(obj.inputs)) {
    throw new ContractParseError(
      'contract.inputs must be "prompt" or a JSON Schema object'
    );
  }

  // outputs (required)
  if (!("outputs" in obj)) {
    throw new ContractParseError("contract.outputs is required");
  }
  if (!isInputOutput(obj.outputs)) {
    throw new ContractParseError(
      'contract.outputs must be "prompt" or a JSON Schema object'
    );
  }

  // permissions (default [])
  const rawPermissions = obj.permissions;
  let permissions: string[] = [];
  if (rawPermissions !== undefined) {
    if (!Array.isArray(rawPermissions)) {
      throw new ContractParseError("contract.permissions must be an array");
    }
    for (const p of rawPermissions) {
      if (typeof p !== "string") {
        throw new ContractParseError("each permission must be a string");
      }
    }
    permissions = rawPermissions as string[];
  }

  // determinism (default "probabilistic")
  let determinism: Determinism = "probabilistic";
  if (obj.determinism !== undefined) {
    if (typeof obj.determinism !== "string" || !VALID_DETERMINISM.has(obj.determinism)) {
      throw new ContractParseError(
        `contract.determinism must be one of: deterministic, bounded, probabilistic — got "${obj.determinism}"`
      );
    }
    determinism = obj.determinism as Determinism;
  }

  // errors (optional)
  let errors: ContractError[] | undefined;
  if (obj.errors !== undefined) {
    if (!Array.isArray(obj.errors)) {
      throw new ContractParseError("contract.errors must be an array");
    }
    errors = [];
    for (const e of obj.errors) {
      if (typeof e !== "object" || e === null) {
        throw new ContractParseError("each contract error must be an object");
      }
      const errObj = e as Record<string, unknown>;
      if (typeof errObj.code !== "string" || typeof errObj.description !== "string") {
        throw new ContractParseError(
          "each contract error must have 'code' and 'description' strings"
        );
      }
      errors.push({ code: errObj.code, description: errObj.description });
    }
  }

  // behavior (optional)
  let behavior: BehavioralMetadata | undefined;
  if (obj.behavior !== undefined) {
    if (typeof obj.behavior !== "object" || obj.behavior === null) {
      throw new ContractParseError("contract.behavior must be an object");
    }
    const bObj = obj.behavior as Record<string, unknown>;
    behavior = {};

    if (bObj.latencyClass !== undefined) {
      if (typeof bObj.latencyClass !== "string" || !VALID_LATENCY.has(bObj.latencyClass)) {
        throw new ContractParseError(
          `behavior.latencyClass must be one of: fast, moderate, slow — got "${bObj.latencyClass}"`
        );
      }
      behavior.latencyClass = bObj.latencyClass as LatencyClass;
    }

    if (bObj.costClass !== undefined) {
      if (typeof bObj.costClass !== "string" || !VALID_COST.has(bObj.costClass)) {
        throw new ContractParseError(
          `behavior.costClass must be one of: free, low, medium, high — got "${bObj.costClass}"`
        );
      }
      behavior.costClass = bObj.costClass as CostClass;
    }

    if (bObj.sideEffects !== undefined) {
      if (typeof bObj.sideEffects !== "boolean") {
        throw new ContractParseError("behavior.sideEffects must be a boolean");
      }
      behavior.sideEffects = bObj.sideEffects;
    }
  }

  return {
    version: obj.version,
    inputs: obj.inputs as InputOutput,
    outputs: obj.outputs as InputOutput,
    permissions,
    determinism,
    errors,
    behavior,
  };
}

/**
 * Validate a parsed contract, returning errors and warnings.
 * Matches the validation logic in the Rust crate.
 */
export function validateContract(contract: Contract): ContractValidationResult {
  const errors: string[] = [];
  const warnings: string[] = [];

  // version must be semver
  if (!isSemver(contract.version)) {
    errors.push(`contract version must be valid semver, got "${contract.version}"`);
  }

  // validate permissions against vocabulary
  for (const perm of contract.permissions) {
    const check = validatePermission(perm);
    if (!check.validFormat) {
      errors.push(`permission "${perm}" has invalid format`);
    } else if (!check.known) {
      warnings.push(`permission "${perm}" is not in the v0 vocabulary`);
    }
  }

  return {
    valid: errors.length === 0,
    errors,
    warnings,
  };
}

File: crates/aule-schema/src/contract.rs
use serde::{Deserialize, Serialize};
use serde::de::Error as _;
use std::path::Path;
use thiserror::Error;

use crate::permissions;
use crate::validation::ValidationResult;

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct Contract {
    pub version: String,
    pub inputs: InputOutput,
    pub outputs: InputOutput,
    #[serde(default)]
    pub permissions: Vec<String>,
    #[serde(default = "default_determinism")]
    pub determinism: Determinism,
    #[serde(default)]
    pub errors: Option<Vec<ContractError>>,
    #[serde(default)]
    pub behavior: Option<BehavioralMetadata>,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(untagged)]
pub enum InputOutput {
    Prompt(PromptMarker),
    Schema(serde_json::Value),
}

// Serde trick: "prompt" string deserializes into this
#[derive(Debug, Clone, Serialize, Deserialize)]
pub enum PromptMarker {
    #[serde(rename = "prompt")]
    Prompt,
}

impl InputOutput {
    pub fn is_prompt(&self) -> bool {
        matches!(self, InputOutput::Prompt(_))
    }
}

#[derive(Debug, Clone, Serialize, Deserialize, PartialEq)]
#[serde(rename_all = "lowercase")]
pub enum Determinism {
    Deterministic,
    Bounded,
    Probabilistic,
}

fn default_determinism() -> Determinism {
    Determinism::Probabilistic
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct ContractError {
    pub code: String,
    pub description: String,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct BehavioralMetadata {
    #[serde(default)]
    pub latency_class: Option<LatencyClass>,
    #[serde(default)]
    pub cost_class: Option<CostClass>,
    #[serde(default)]
    pub side_effects: Option<bool>,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "lowercase")]
pub enum LatencyClass {
    Fast,
    Moderate,
    Slow,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "lowercase")]
pub enum CostClass {
    Free,
    Low,
    Medium,
    High,
}

// --- Errors ---

#[derive(Debug, Error)]
pub enum ContractParseError {
    #[error("YAML parse error: {0}")]
    YamlParse(#[from] serde_yaml::Error),
    #[error("contract file not found: {0}")]
    NotFound(String),
}

// --- Source for parsing ---

pub enum ContractSource<'a> {
    Inline(&'a serde_json::Value),
    File(&'a Path),
}

// --- Parsing ---

pub fn parse_contract_from_yaml(yaml: &str) -> Result<Contract, ContractParseError> {
    let contract: Contract = serde_yaml::from_str(yaml)?;
    Ok(contract)
}

pub fn parse_contract(source: ContractSource) -> Result<Contract, ContractParseError> {
    match source {
        ContractSource::Inline(value) => {
            let contract: Contract = serde_json::from_value(value.clone())
                .map_err(|e| serde_yaml::Error::custom(e.to_string()))?;
            Ok(contract)
        }
        ContractSource::File(path) => {
            let content = std::fs::read_to_string(path)
                .map_err(|_| ContractParseError::NotFound(path.display().to_string()))?;
            parse_contract_from_yaml(&content)
        }
    }
}

// --- Validation ---

pub fn validate_contract(contract: &Contract) -> ValidationResult {
    let mut result = ValidationResult::new();

    // version must be semver
    if contract.version.split('.').count() != 3
        || contract
            .version
            .split('.')
            .any(|p| p.parse::<u64>().is_err())
    {
        result.add_error(format!(
            "contract version must be valid semver, got \"{}\"",
            contract.version
        ));
    }

    // validate permissions against vocabulary
    for perm in &contract.permissions {
        let check = permissions::validate_permission(perm);
        if !check.valid_format {
            result.add_error(format!("permission \"{}\" has invalid format", perm));
        } else if !check.known {
            result.add_warning(format!(
                "permission \"{}\" is not in the v0 vocabulary",
                perm
            ));
        }
    }

    result
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn parse_prompt_based_contract() {
        let yaml = r#"
version: "1.0.0"
inputs: "prompt"
outputs: "prompt"
permissions: []
determinism: "probabilistic"
"#;
        let contract = parse_contract_from_yaml(yaml).unwrap();
        assert!(contract.inputs.is_prompt());
        assert!(contract.outputs.is_prompt());
        assert_eq!(contract.determinism, Determinism::Probabilistic);
    }

    #[test]
    fn parse_structured_contract() {
        let yaml = r#"
version: "1.0.0"
inputs:
  type: "object"
  properties:
    query:
      type: "string"
  required: ["query"]
outputs: "prompt"
permissions: ["filesystem.read"]
determinism: "bounded"
"#;
        let contract = parse_contract_from_yaml(yaml).unwrap();
        assert!(!contract.inputs.is_prompt());
        assert!(contract.outputs.is_prompt());
        assert_eq!(contract.determinism, Determinism::Bounded);
    }

    #[test]
    fn default_determinism_is_probabilistic() {
        let yaml = r#"
version: "1.0.0"
inputs: "prompt"
outputs: "prompt"
permissions: []
"#;
        let contract = parse_contract_from_yaml(yaml).unwrap();
        assert_eq!(contract.determinism, Determinism::Probabilistic);
    }

    #[test]
    fn validate_unknown_permission_warns() {
        let yaml = r#"
version: "1.0.0"
inputs: "prompt"
outputs: "prompt"
permissions: ["quantum.entangle"]
"#;
        let contract = parse_contract_from_yaml(yaml).unwrap();
        let result = validate_contract(&contract);
        assert!(result.is_ok()); // warnings only
        assert!(!result.warnings().is_empty());
    }

    #[test]
    fn validate_bad_version() {
        let yaml = r#"
version: "not-semver"
inputs: "prompt"
outputs: "prompt"
permissions: []
"#;
        let contract = parse_contract_from_yaml(yaml).unwrap();
        let result = validate_contract(&contract);
        assert!(!result.is_ok());
    }

    #[test]
    fn parse_with_behavioral_metadata() {
        let yaml = r#"
version: "1.0.0"
inputs: "prompt"
outputs: "prompt"
permissions: []
behavior:
  latencyClass: "slow"
  sideEffects: true
"#;
        let contract = parse_contract_from_yaml(yaml).unwrap();
        let behavior = contract.behavior.unwrap();
        assert!(matches!(behavior.latency_class, Some(LatencyClass::Slow)));
        assert_eq!(behavior.side_effects, Some(true));
    }

    #[test]
    fn parse_with_custom_errors() {
        let yaml = r#"
version: "1.0.0"
inputs: "prompt"
outputs: "prompt"
permissions: []
errors:
  - code: "CONTEXT_TOO_LARGE"
    description: "Input exceeds context window"
"#;
        let contract = parse_contract_from_yaml(yaml).unwrap();
        let errors = contract.errors.unwrap();
        assert_eq!(errors.len(), 1);
        assert_eq!(errors[0].code, "CONTEXT_TOO_LARGE");
    }
}
